File: AUGUST-2025/11-08-2025/p2438.dart
class Solution {
    List<int> productQueries(int n, List<List<int>> queries) {
        int mod = pow(10, 9).toInt() + 7;
        List<int> powers = [];
        List<int> valid = [];
        for (int i = 0; i < 32; i++) {
            powers.add(pow(2, i).toInt());
        }

        var index = powers.lastIndexWhere((e) => e <= n);

        while (n > 0) {
            valid.add(powers[index]);
            n -= powers[index];
            index = powers.lastIndexWhere((e) => e <= n);
        }

        List<int> p = valid.reversed.toList();
        List<int> res = [];
        List<BigInt> prefix = List.filled(p.length, BigInt.zero);
        prefix[0] = BigInt.from(p[0]);

        for (int i = 1; i < p.length; i++) {
            prefix[i] = BigInt.from(p[i]) * prefix[i - 1];
        }

        for (final query in queries) {
            final start = query[0];
            final end = query[1];
            if (start == 0) {
                res.add((prefix[end] % BigInt.from(mod)).toInt());
            } else {
                res.add(((prefix[end] ~/ prefix[start - 1]) % BigInt.from(mod)).toInt());
            }
        }

        return res;
    }
}

File: AUGUST-2025/11-08-2025/p2438.cpp
class Solution {
public:
    vector<int>powersOf2(int n){
        vector<int>v;

        int curr = 1;
        while(n > 0){
            if(n % 2 == 1){
                v.push_back(curr);
            }
            curr *= 2;
            n /= 2;
        }
        return v;
    }

    vector<int> productQueries(int n, vector<vector<int>>& queries) {
        const int MOD = 1e9 + 7;
        vector<int>res;
        vector<int>powers = powersOf2(n);
        
        for(auto it : queries){
            long long prod = 1;
            for(int i = it[0]; i <= it[1]; i++){
                prod = (prod * powers[i]) % MOD;
            }
            res.push_back(prod);
        }
        
        return res;
    }
};